psydac/core/bsplines.py

```diff
@@ -57,7 +57,7 @@
 
 
 #==============================================================================
-def find_span(knots, degree, x):
+def find_span(knots, degree, x, periodic, multiplicity = 1):
```

Comment:
The docstring should be a bit more precise:

here I guess that the `knots` passed as argument are assumed to have all the same `multiplicity`, except for boundary knots when `periodic = False`, which always have multiplicity `degree+1` . Is that correct ?

and the actual knot sequence which will be used 

here I assume that there is no repea

you mean: multiplicity of the interior knots (since boundary knots always have multiplicity = degree+1 in the non-periodic case)

Reply:
Yes, I'll add it to the doc

Reply:
Thanks. Also in the periodic case, it would help to specify (remind ?) what is actually in the knot vector -- namely if some knots are "repeated" outside the domain to have an efficient handling of the periodicity condition.

Reply:
Maybe I should just say that we assume that the knots sequence was created by `make_knots` and then in `make_knots` specify what it looks like?

Reply:
yes, that's a good idea -- I was actually missing this pointer to `make_knots`

and by doing this you can improve the docstring in `make_knots`: for instance the sentence 
```
Otherwise, knot sequence is clamped (i.e. endpoints are repeated p times).
```
is ambiguous (what does it mean to be repeated "one time" ?), it would be better to write:
```
Otherwise, knot sequence is clamped (i.e. endpoints have p+1 multiplicity).
```

Reply:
I do not understand why we should pass `multiplicity` to this function. The knot sequence already contains all the required information, and the computed `span` is the index of the last non-zero basis function, which has the same meaning regardless of the multiplicity.

Reply:
I was also a bit confused at the start -- thinking first that passing some `multiplicity` > 1 would further repeat the given knots, which is not the case

Reply:
I see that the interior `multiplicity` is used to compute the `high` index in the periodic case. Do we get the wrong `x` coordinate of the right boundary otherwise?

Reply:
In the periodic case `high` would not be right, however `low` remains untouched this is due to how I compute the knots in the periodic case, for example if our space is `[a,b]` with `n` cells, my knots will be `a+i/n*(b-a)` for `i `in `[1,n]`
Then I will add `p+1` knots on the left by translating the rightmost knots, meaning that `low` is still `p`, 
But I'll add `p` more knots to the right and this is why we need to change `high` in this case

Reply:
Could we pass the dimension of the spline space instead of `multiplicity`?

Reply:
Well since the dimension of the space, the number of knots the multiplicity and the degree are all linked we could use that yes but is it really better? I don't see why passing the multiplicity is any bad here?

Reply:
I find that using the multiplicity is not satisfactory because I would like this function to work for any knot sequence, where the multiplicity can vary arbitrarily from one knot to another. This is something that we want to support in the future.

Reply:
Conceptually I think that the knot sequence should be extended to the left always in the same way. As a result, the basis function of index 0 is the left-most B-spline which is non-zero on the left boundary.

Reply:
That is also what I would think -- unless I am missing something ?

Reply:
If I am not mistaken this is the case with how things are build at the moment?

Reply:
What we could do is : 
-Build the knot sequence by repeating all the break mult times (including the boundary ones blindly).
-Add (p-m+1) knots at each side by periodicity.

Then we would have the same number of knots as in the None periodic case and enough knots to evaluate everything inside the domain.

Reply:
Yes this is the same. This would not cause problem with the ghost regions then?

Reply:
Yes. Also, note that in this way we have the same number of knots in the periodic and non-periodic case. The only difference is that in the non-periodic case the "extended" knots are all on the boundaries (p+1 on each side)

Reply:
and if I'm not mistaken you no longer need to pass any multiplicity argument to the `find_span` function